include/unifex/indexed_for.hpp

```diff
@@ -151,29 +318,58 @@ struct indexed_for_sender {
     template <typename Visit>
     friend void tag_invoke(
         tag_t<visit_continuations>,
-        const indexed_for_receiver& r,
+        const double_indexed_for_receiver& r,
         Visit&& visit) {
       std::invoke(visit, r.receiver_);
     }
+
   };
 
   template <typename Receiver>
   auto connect(Receiver&& receiver) && {
+    // Connect the receiver to the predecssor of pred_
+    // we skip the connect/start pairing on the immediate
+    // predecessor as an optimisation (or at least, to
+    // emulate the optimisation we might see in
+    // accelerator implementations)
     return unifex::connect(
-        std::forward<Predecessor>(pred_),
-        indexed_for_receiver<std::remove_cvref_t<Receiver>>{
+        std::forward<Predecessor>(pred_).pred_,
+        double_indexed_for_receiver<std::remove_cvref_t<Receiver>>{
+            std::forward<Predecessor>(pred_).func_,
             std::forward<Func>(func_),
+            std::forward<Predecessor>(pred_).policy_,
             std::forward<Policy>(policy_),
+            std::forward<Predecessor>(pred_).range_or_selector_,
             std::forward<RangeOrRangeSelector>(range_or_selector_),
             std::forward<Receiver>(receiver)});
   }
 };
 
-template <typename Sender, typename Policy, typename RangeOrRangeSelector, typename Func>
-auto indexed_for(Sender&& predecessor, Policy&& policy, RangeOrRangeSelector&& range, Func&& func) {
-  return indexed_for_sender<
-      std::remove_cvref_t<Sender>, std::decay_t<Policy>, std::decay_t<RangeOrRangeSelector>, std::decay_t<Func>>{
-      (Sender &&) predecessor, (Policy&&) policy, (RangeOrRangeSelector&& ) range, (Func &&) func};
+// Customisation for when two indexed_for_senders are chained
+// This simulates how we might customise an arbitrary set of types for
+// optimal interaction
+template <typename InnerPredecessor, typename InnerPolicy, typename InnerRange, typename InnerFunc, typename Policy, typename Range, typename Func>
+auto
+tag_invoke(
```

Comment:
We should try to find a way to avoid putting this definition at namespace scope where it could be added to the overload-set for every invocation of every unifex CPO.

One idea is to define the `indexed_for_cpo` type in a nested cpo-specific namespace and then put the overload in there. Other ways are possible by declaring things in a different order.

Reply:
I put all of this in a detail namespace as I was tidying the CPO anyway.

Maybe we should define a namespace for each algorithm with a known name, and that's where you are supposed to put tag_invoke customisations?